File: 22-GenerateParentheses/22-GenerateParentheses.java
// Last updated: 6/10/2026, 7:28:03 PM
class Solution {
    private void backtrack(List<String> res, StringBuilder cur, int open, int close, int n) {
        // Base case: string is complete
        if (cur.length() == 2 * n) {
            res.add(cur.toString());
            return;
        }

        // Try adding '(' if we still can
        if (open < n) {
            cur.append("(");
            backtrack(res, cur, open + 1, close, n);
            cur.deleteCharAt(cur.length() - 1); // backtrack
        }

        // Try adding ')' if valid
        if (close < open) {
            cur.append(")");
            backtrack(res, cur, open, close + 1, n);
            cur.deleteCharAt(cur.length() - 1); // backtrack
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        backtrack(res, new StringBuilder(), 0, 0, n);
        return res;
    }

}

File: 22-GenerateParentheses/22-GenerateParentheses.cpp
// Last updated: 6/10/2026, 7:27:59 PM
class Solution {
public:
    int total;
    vector<string> ans;
    vector<string> generateParenthesis(int n) {
        total = n;
        string curr = "";
        gen(0 , 0 , curr);
        return ans;
    }
private:
    void gen(int open , int cnt , string curr) {
        if (curr.size() == 2 * total) {
            ans.push_back(curr);
            return;
        }
        if (open) {
            curr.push_back(')');
            gen(open-1 , cnt , curr);
            curr.pop_back();
        }
        if (cnt < total) {
            curr.push_back('(');
            gen(open+1 , cnt+1 , curr);
            curr.pop_back();
        }
    }
};